cpp/src/strings/convert/convert_urls.cu

```diff
@@ -328,82 +396,54 @@ std::unique_ptr<column> url_decode(
-                                         thrust::make_counting_iterator<size_type>(chars_end),
-                                         esc_detector);
-
-  if (esc_count == 0) {
-    // nothing to replace, so just copy the input column
-    return std::make_unique<cudf::column>(strings.parent(), stream, mr);
-  }
-
-  // create a vector of the potential escape sequence positions
-  rmm::device_uvector<size_type> esc_positions(esc_count, stream);
-  auto d_esc_positions = esc_positions.data();
-  thrust::copy_if(rmm::exec_policy(stream),
-                  thrust::make_counting_iterator<size_t>(chars_start),
-                  thrust::make_counting_iterator<size_t>(chars_end),
-                  d_esc_positions,
-                  esc_detector);
-
-  // In-place remove any escape positions that crossed string boundaries.
-  device_span<int32_t const> d_offsets_span(d_offsets, offset_count);
-  auto esc_pos_end = thrust::remove_if(rmm::exec_policy(stream),
-                                       d_esc_positions,
-                                       d_esc_positions + esc_count,
-                                       url_decode_esc_position_filter{d_offsets_span});
-
-  // update count in case any were filtered
-  esc_count = esc_pos_end - d_esc_positions;
-  if (esc_count == 0) {
-    // nothing to replace, so just copy the input column
-    return std::make_unique<cudf::column>(strings.parent(), stream, mr);
-  }
-
-  device_span<size_type const> d_esc_positions_span(d_esc_positions, esc_count);
 
   // build offsets column
   auto offsets_column = make_numeric_column(
     data_type{type_id::INT32}, offset_count, mask_state::UNALLOCATED, stream, mr);
-  auto offsets_view = offsets_column->mutable_view();
-  thrust::transform(rmm::exec_policy(stream),
-                    d_offsets_span.begin(),
-                    d_offsets_span.end(),
-                    offsets_view.begin<int32_t>(),
-                    url_decode_offsets_updater{d_esc_positions_span, chars_start});
+
+  // count number of bytes in each string after decoding and store it in offsets_column
+  auto offsets_view         = offsets_column->view();
+  auto offsets_mutable_view = offsets_column->mutable_view();
+  url_decode_char_counter<num_warps_per_threadblock, char_block_size>
+    <<<num_threadblocks, threadblock_size, 0, stream.value()>>>(
+      d_in_chars, d_in_offsets, offsets_mutable_view.begin<offset_type>() + 1, strings_count);
+
+  // use inclusive_scan to transform number of bytes into offsets
+  thrust::inclusive_scan(rmm::exec_policy(stream),
+                         offsets_view.begin<offset_type>() + 1,
+                         offsets_view.begin<offset_type>() + offset_count,
+                         offsets_mutable_view.begin<offset_type>() + 1);
+
+  // set the first element of the offset column to 0
+  CUDA_TRY(cudaMemsetAsync(
+    offsets_mutable_view.begin<offset_type>(), 0, sizeof(offset_type), stream.value()));
```

Comment:
If you stored the sizes in the offsets column starting at position 0 (instead of position 1) and then do an exclusive-scan, I don't think you will need the `cudaMemset` call
```suggestion
  url_decode_char_counter<num_warps_per_threadblock, char_block_size>
    <<<num_threadblocks, threadblock_size, 0, stream.value()>>>(
      d_in_chars, d_in_offsets, offsets_mutable_view.begin<offset_type>(), strings_count);

  // use scan to transform number of bytes into offsets
  thrust::exclusive_scan(rmm::exec_policy(stream),
                         offsets_view.begin<offset_type>(),
                         offsets_view.end<offset_type>(),
                         offsets_mutable_view.begin<offset_type>());
```

Reply:
Implemented in https://github.com/rapidsai/cudf/pull/8622/commits/2964dedf6671d088c0238f09fcc35a26a41339ce.